### ISSUE TITLE
rpmdb2solv is broken with rpm 4.16 on Fedora

### ISSUE THREAD

**Issue by Conan-Kudo** [openSUSE/libsolv#412]
It seems that `rpmdb2solv` doesn't work properly with RPM 4.16 on Fedora. I'm seeing failures in the Zypper test suite on Fedora 33+ like so:

```
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.KzTUmx
+ umask 022
+ cd /builddir/build/BUILD
+ cd zypper-1.14.37
+ pushd x86_64-redhat-linux-gnu/tests
~/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests ~/build/BUILD/zypper-1.14.37
+ ctest -VV --output-on-failure .
UpdateCTestConfiguration  from :/builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/DartConfiguration.tcl
UpdateCTestConfiguration  from :/builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/DartConfiguration.tcl
Test project /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: PackageArgs_test
1: Test command: /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/PackageArgs_test "--catch_system_errors=no"
1: Test timeout computed to be: 10000000
1: Running 6 test cases...
1: 
1: [1;32;49m*** No errors detected
1: [0;39;49m
1/6 Test #1: PackageArgs_test .................   Passed    0.02 sec
test 2
    Start 2: SolverRequester_test
2: Test command: /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/SolverRequester_test "--catch_system_errors=no"
2: Test timeout computed to be: 10000000
2: Running 36 test cases...
2: 'info-4.13-1.1' not found in package names. Trying 'info = 4.13-1.1'.
2: 'openSUSE-release-usb-x11' not found in package names. Trying 'openSUSE-release = usb-x11'.
2: 'cron-4.1-195.0.x86_64' not found in package names. Trying 'cron.x86_64 = 4.1-195.0'.
2: 'patch:libxml2-434' not found in package names. Trying 'patch:libxml2 = 434'.
2: 
2: [1;32;49m*** No errors detected
2: [0;39;49m
2/6 Test #2: SolverRequester_test .............   Passed    1.31 sec
test 3
    Start 3: ZyppFlags_test
3: Test command: /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/ZyppFlags_test "--catch_system_errors=no"
3: Test timeout computed to be: 10000000
3: Running 7 test cases...
3: 
3: [1;32;49m*** No errors detected
3: [0;39;49m
3/6 Test #3: ZyppFlags_test ...................   Passed    0.02 sec
test 4
    Start 4: Locales_test
4: Test command: /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/Locales_test "--catch_system_errors=no"
4: Test timeout computed to be: 10000000
4: Initializing Target
4: Target initialization failed:
4: Failed to cache rpm database (1).
4: History:
4:  - 'rpmdb2solv' '-r' '/var/tmp/TmpDir.NxHkGc' '-X' '-p' '/var/tmp/TmpDir.NxHkGc/etc/products.d' '-o' '/var/tmp/TmpDir.NxHkGc/var/cache/zypp/solv/@System/solvrmXXQ9'
4: warning: invalid %_db_backend: 
4: warning: using dummy database, installs not possible
4: rpmdb2solv: no error
4: 
4: 
4: Running 'zypper refresh' as root might resolve the problem.
4: unknown location(0): fatal error: in "Test setup": ExitRequestException: Target initialization failed: Failed to cache rpm database (1).
4: Running 2 test cases...
4: 
4: [1;32;49m*** No errors detected
4: [0;39;49mTest setup error: 
4/6 Test #4: Locales_test .....................***Failed    0.32 sec
Initializing Target
Target initialization failed:
Failed to cache rpm database (1).
History:
 - 'rpmdb2solv' '-r' '/var/tmp/TmpDir.NxHkGc' '-X' '-p' '/var/tmp/TmpDir.NxHkGc/etc/products.d' '-o' '/var/tmp/TmpDir.NxHkGc/var/cache/zypp/solv/@System/solvrmXXQ9'
warning: invalid %_db_backend: 
warning: using dummy database, installs not possible
rpmdb2solv: no error
Running 'zypper refresh' as root might resolve the problem.
unknown location(0): fatal error: in "Test setup": ExitRequestException: Target initialization failed: Failed to cache rpm database (1).
Running 2 test cases...
[1;32;49m*** No errors detected
[0;39;49mTest setup error: 
test 5
    Start 5: text_test
5: Test command: /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/utils/text_test "--catch_system_errors=no"
5: Test timeout computed to be: 10000000
5: Running 5 test cases...
5: locale set to: en_US.UTF-8
5: locale set to: en_US.UTF-8
5: 
5: [1;32;49m*** No errors detected
5: [0;39;49m
5/6 Test #5: text_test ........................   Passed    0.02 sec
test 6
    Start 6: formater_test
6: Test command: /builddir/build/BUILD/zypper-1.14.37/x86_64-redhat-linux-gnu/tests/utils/formater_test "--catch_system_errors=no"
6: Test timeout computed to be: 10000000
6: Running 2 test cases...
6: 
6: [1;32;49m*** No errors detected
6: [0;39;49m
6/6 Test #6: formater_test ....................   Passed    0.01 sec
83% tests passed, 1 tests failed out of 6
Total Test time (real) =   1.70 sec
The following tests FAILED:
	  4 - Locales_test (Failed)
Errors while running CTest
```

Full build log: [zypper-1.14.37-1.fc34-build.log.txt](https://github.com/openSUSE/libsolv/files/5579864/zypper-1.14.37-1.fc34-build.log.txt)

Here's a recent Koji task showing the issue: https://koji.fedoraproject.org/koji/taskinfo?taskID=55578877

This is also tracked as [`rh#1879952`](https://bugzilla.redhat.com/show_bug.cgi?id=1879952).

**Comment by Conan-Kudo** [comment]
This is caused by https://github.com/rpm-software-management/rpm/commit/ac68897ddd077e78766b73bd837eabbeab0b242d

**Comment by mlschroe** [comment]
Yes, known issue. See my comment in https://github.com/rpm-software-management/rpm/commit/471b7be4bd5cc7f245f9aa00c7784a7056e439b7

**Comment by mlschroe** [comment]
Unfortunately I didn't catch that before 4.16.0 was released.

(openSUSE has a patch in its rpm to make it work again.)

**Comment by Conan-Kudo** [comment]
Wouldn't the correct fix be to pull the `%_db_backend` setting and properly do error handling if no database exists?

**Comment by mlschroe** [comment]
You mean in the rpm code?

**Comment by Conan-Kudo** [comment]
I mean in `rpmdb2solv`. We should probably be pulling the setting from RPM and using it the same time we set the `%_dbpath` macro. I would imagine this problem goes away if you just set `%_db_backend` to the same value that the system RPM has, right?

**Comment by mlschroe** [comment]
But librpm doesn't return an error, it just returns no result. There's nothing I can do in libsolv to detect this.

**Comment by Conan-Kudo** [comment]
Sure, but when I patch rpmdb2solv with a value for `%_db_backend` everything works. I'd think the correct solution here would be to make sure we pull the value from RPM and use it when we try to do stuff with the rpmdb.

**Comment by mlschroe** [comment]
Fixed with commit c78a330a27bb3844b97e694552efc4a373cf217b
